mlflow/utils/gorilla.py

```diff
@@ -313,27 +314,16 @@ def apply(patch):
         if settings.store_hit:
             original_name = _ORIGINAL_NAME % (patch.name,)
             curr_active_patch = _ACTIVE_PATCH % (patch.name,)
-            # For certain MLflow Models use cases, such as scikit-learn autologging, we patch
-            # a method on a parent class
-            # (e.g., `sklearn.feature_extraction.text.CountVectorizer.fit_transform()`) and
-            # subsequently patch a corresponding overridden method on one of its children
-            # (e.g., `feature_extraction.text.TfidfVectorizer.fit_transform()`)
-            # to provide a different implementation. In these cases, we wish to swap out the
-            # "original function" attribute stored on the child in order to refer to the child's
-            # overriden method (e.g., `feature_extraction.text.TfidfVectorizer.fit_transform()`),
-            # rather than the parent method
-            # (e.g., `sklearn.feature_extraction.text.CountVectorizer.fit_transform()`)
-            prev_patch = getattr(patch.destination, curr_active_patch, None)
-
-            if not hasattr(patch.destination, original_name) or (
-                prev_patch
-                and prev_patch.destination != patch.destination
-                and issubclass(patch.destination, prev_patch.destination)
-            ):
-                setattr(patch.destination, original_name, target)
-
-    setattr(patch.destination, patch.name, patch.obj)
-    setattr(patch.destination, curr_active_patch, patch)
+            # Note: the patching rule is, we only patch the destination class,
+            #   it don't modify the parent class. The following `setattr` will only add attribute
+            #   into destination class.
+            if original_name in patch.destination.__dict__:
+                raise RuntimeError(
+                    f'Patch for {patch.destination.__name__}.{patch.name} already exists.')
+
+            setattr(patch.destination, original_name, target)
+            setattr(patch.destination, patch.name, patch.obj)
+            setattr(patch.destination, curr_active_patch, patch)
```

Comment:
I simplify the logic here:

(1) We always patch in destination class, even if the attribute to be patched exist in parent class. This rule address the sub-class overridden method issue.
(2) I add a strict checking `if original_name in patch.destination.__dict__` then raise error, because when setup autologging each method should only be patch once, if patch again on a patched method then it means code somewhere is wrong.